src/main/java/org/opensearch/neuralsearch/query/SparseEncodingQueryBuilder.java

```diff
@@ -0,0 +1,306 @@
+                );
+            }
+        }
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        // If the user has specified query_tokens field, then we don't need to inference the sentence,
+        // just re-rewrite to self. Otherwise, we need to inference the sentence to get the queryTokens. Then the
+        // logic is similar to NeuralQueryBuilder
+        if (null != queryTokens) {
+            return this;
+        }
+        if (null != queryTokensSupplier) {
+            return queryTokensSupplier.get() == null
+                ? this
+                : new SparseEncodingQueryBuilder().fieldName(fieldName)
+                    .queryTokens(queryTokensSupplier.get())
+                    .queryText(queryText)
+                    .modelId(modelId);
+        }
+
+        validateForRewrite(queryText, modelId);
+        SetOnce<Map<String, Float>> queryTokensSetOnce = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction(
+            ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
+                modelId(),
+                List.of(queryText),
+                ActionListener.wrap(mapResultList -> {
+                    queryTokensSetOnce.set(TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0));
+                    actionListener.onResponse(null);
+                }, actionListener::onFailure)
+            ))
+        );
+        return new SparseEncodingQueryBuilder().fieldName(fieldName)
+            .queryText(queryText)
+            .modelId(modelId)
+            .queryTokensSupplier(queryTokensSetOnce::get);
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        final MappedFieldType ft = context.fieldMapper(fieldName);
+        validateFieldType(ft);
+        validateQueryTokens(queryTokens);
+
+        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+        for (Map.Entry<String, Float> entry : queryTokens.entrySet()) {
+            builder.add(FeatureField.newLinearQuery(fieldName, entry.getKey(), entry.getValue()), BooleanClause.Occur.SHOULD);
+        }
+        return builder.build();
+    }
+
+    private static void validateForRewrite(String queryText, String modelId) {
+        if (null == queryText || null == modelId) {
+            throw new IllegalArgumentException(
+                "When "
+                    + QUERY_TOKENS_FIELD.getPreferredName()
+                    + " are not provided,"
+                    + QUERY_TEXT_FIELD.getPreferredName()
+                    + " and "
+                    + MODEL_ID_FIELD.getPreferredName()
+                    + " can not be null."
+            );
+        }
+    }
+
+    private static void validateFieldType(MappedFieldType fieldType) {
+        if (!fieldType.typeName().equals("rank_features")) {
```

Comment:
Is it possible the fieldType.typeName() return null?

Reply:
This is a good point. The fieldType can be null, and I missed this check, will add one. However the fieldType.typeName() is an abstract method in the super class, and the inheritors need to implement it explicitly. I sampled about 10 inheritors and all of them hard-code the return value or return a static String field of the inheritor. So I don't think fieldType.typeName() can return null.